ucp/benchmarks/backends/ucp_core.py

```diff
@@ -0,0 +1,287 @@
+
+        # +1 to account for wireup message
+        if self.args.delay_progress:
+            while finished[0] < self.args.n_iter + self.args.n_warmup_iter + 1 and (
+                outstanding[0] >= self.args.max_outstanding
+                or finished[0] + self.args.max_outstanding
+                >= self.args.n_iter + self.args.n_warmup_iter + 1
+            ):
+                worker.progress()
+        else:
+            while finished[0] != self.args.n_iter + self.args.n_warmup_iter + 1:
+                worker.progress()
+
+
+class UCXPyCoreClient(BaseClient):
+    def __init__(self, args, xp, queue, server_address, port):
+        self.args = args
+        self.xp = xp
+        self.queue = queue
+        self.server_address = server_address
+        self.port = port
+
+    def run(self):
+        ctx = ucx_api.UCXContext(
+            feature_flags=(
+                ucx_api.Feature.AM
+                if self.args.enable_am is True
+                else ucx_api.Feature.TAG,
+            )
+        )
+        worker = ucx_api.UCXWorker(ctx)
+        register_am_allocators(self.args, worker)
+        ep = ucx_api.UCXEndpoint.create(
+            worker,
+            self.server_address,
+            self.port,
+            endpoint_error_handling=True,
+        )
+
+        send_msg = self.xp.arange(self.args.n_bytes, dtype="u1")
+        if self.args.reuse_alloc:
+            recv_msg = self.xp.zeros(self.args.n_bytes, dtype="u1")
+
+        if self.args.enable_am:
+            blocking_am_send(worker, ep, send_msg)
+            blocking_am_recv(worker, ep)
+        else:
+            wireup_recv = bytearray(len(WireupMessage))
+            blocking_send(worker, ep, WireupMessage)
+            blocking_recv(worker, ep, wireup_recv)
+
+        op_lock = Lock()
+        finished = [0]
+        outstanding = [0]
+
+        def maybe_progress():
+            while outstanding[0] >= self.args.max_outstanding:
+                worker.progress()
+
+        def op_started():
+            with op_lock:
+                outstanding[0] += 1
+
+        def op_completed():
+            with op_lock:
+                outstanding[0] -= 1
+                finished[0] += 1
+
+        if self.args.cuda_profile:
+            self.xp.cuda.profiler.start()
+
+        times = []
+        for i in range(self.args.n_iter + self.args.n_warmup_iter):
+            start = monotonic()
+
+            if self.args.enable_am:
+                blocking_am_send(worker, ep, send_msg)
+                blocking_am_recv(worker, ep)
+            else:
+                if not self.args.reuse_alloc:
+                    recv_msg = self.xp.zeros(self.args.n_bytes, dtype="u1")
+
+                if self.args.delay_progress:
+                    maybe_progress()
+                    non_blocking_send(worker, ep, send_msg, op_started, op_completed)
+                    maybe_progress()
+                    non_blocking_recv(worker, ep, recv_msg, op_started, op_completed)
```

Comment:
Usual convention for non-blocking point to point is to post receives before sends.

Reply:
So:
```
non_blocking_recv(...)
non_blocking_send(...)
maybe_progress()
```
?

Reply:
For the Tag API, that's ok. I'm generally hesitant to make this kind of change because the stream API doesn't guarantee the ordering: https://github.com/rapidsai/ucx-py/blob/branch-0.28/ucp/core.py#L48-L55 .

Reply:
In any case applied that change in https://github.com/rapidsai/ucx-py/pull/873/commits/9a88d1cf5324a9ecce86270eb1c0a6ad384fc418 .